File: src/api/sources.js
import { invoke } from "@tauri-apps/api/core";

/** 书源管理、规则校验、目录/正文解析 */

export async function validateSourceRule(rule) {
  return invoke("validate_source_rule", {
    ruleJson: JSON.stringify(rule),
  });
}

export async function fetchChapters(url, rule) {
  return invoke("fetch_chapters", {
    url,
    ruleJson: JSON.stringify(rule),
  });
}

export async function getBookContent(chapterUrl, rule) {
  return invoke("get_book_content", {
    chapterUrl,
    ruleJson: JSON.stringify(rule),
  });
}

export async function listBookSources() {
  return invoke("list_book_sources");
}

export async function toggleBookSource(sourceId, enabled) {
  return invoke("toggle_book_source", { sourceId, enabled });
}

export async function subscribeRemoteSource(url) {
  return invoke("subscribe_remote_source", { url });
}

export async function importBookSourcesBatch(ruleJson) {
  return invoke("import_book_sources_batch", { ruleJson });
}

export async function importBookSourceJson(ruleJson) {
  return invoke("import_book_source_json", { ruleJson });
}

/** 列出已订阅的书源仓库 */
export async function listSourceSubscriptions() {
  return invoke("list_source_subscriptions");
}

/** 重新同步订阅仓库 */
export async function syncSourceSubscription(subId) {
  return invoke("sync_source_subscription", { subId });
}

/** 探测单书源健康状态 */
export async function pingBookSource(sourceId) {
  return invoke("ping_book_source", { sourceId });
}

/** 批量探测书源 */
export async function pingBookSources(sourceIds) {
  return invoke("ping_book_sources", { sourceIds });
}

/** 批量删除非内置书源 */
export async function deleteBookSources(sourceIds) {
  return invoke("delete_book_sources", { sourceIds });
}

/** 批量启用/禁用书源 */
export async function setBookSourcesEnabled(sourceIds, enabled) {
  return invoke("set_book_sources_enabled", { sourceIds, enabled });
}

File: src-tauri/src/commands/sources.rs
//! 书源管理、规则校验、目录/正文解析、零配置接入

use crate::services::{source, to_cmd_err};
use crate::spider::{AutoConnectResult, DetectLogger, DetectResponse};
use crate::storage::{SourceHealth, SourceRecord, SourceSubscriptionRecord, SubscribeBatchResult};

#[tauri::command]
pub fn validate_source_rule(rule_json: String) -> Result<String, String> {
    source::validate_source_rule(&rule_json).map_err(to_cmd_err)
}

#[tauri::command]
pub fn fetch_chapters(url: String, rule_json: String) -> Result<Vec<crate::spider::Chapter>, String> {
    source::fetch_chapters(&url, &rule_json).map_err(to_cmd_err)
}

#[tauri::command]
pub fn get_book_content(chapter_url: String, rule_json: String) -> Result<String, String> {
    source::fetch_chapter_content(&chapter_url, &rule_json).map_err(to_cmd_err)
}

#[tauri::command]
pub fn list_book_sources() -> Result<Vec<SourceRecord>, String> {
    source::list_book_sources().map_err(to_cmd_err)
}

#[tauri::command]
pub fn toggle_book_source(source_id: String, enabled: bool) -> Result<SourceRecord, String> {
    source::toggle_book_source(&source_id, enabled).map_err(to_cmd_err)
}

#[tauri::command]
pub fn subscribe_remote_source(url: String) -> Result<SubscribeBatchResult, String> {
    source::subscribe_remote_source(&url).map_err(to_cmd_err)
}

#[tauri::command]
pub fn import_book_sources_batch(rule_json: String) -> Result<SubscribeBatchResult, String> {
    source::import_book_sources_batch(&rule_json).map_err(to_cmd_err)
}

#[tauri::command]
pub fn import_book_source_json(rule_json: String) -> Result<SourceRecord, String> {
    source::import_book_source_json(&rule_json).map_err(to_cmd_err)
}

#[tauri::command]
pub fn list_source_subscriptions() -> Result<Vec<SourceSubscriptionRecord>, String> {
    source::list_source_subscriptions().map_err(to_cmd_err)
}

#[tauri::command]
pub fn sync_source_subscription(sub_id: String) -> Result<SubscribeBatchResult, String> {
    source::sync_source_subscription(&sub_id).map_err(to_cmd_err)
}

#[tauri::command]
pub fn ping_book_source(source_id: String) -> Result<SourceHealth, String> {
    source::ping_source(&source_id).map_err(to_cmd_err)
}

#[tauri::command]
pub fn ping_book_sources(source_ids: Vec<String>) -> Result<Vec<SourceHealth>, String> {
    source::ping_sources_batch(source_ids).map_err(to_cmd_err)
}

#[tauri::command]
pub fn delete_book_sources(source_ids: Vec<String>) -> Result<usize, String> {
    source::delete_sources_batch(source_ids).map_err(to_cmd_err)
}

#[tauri::command]
pub fn set_book_sources_enabled(source_ids: Vec<String>, enabled: bool) -> Result<usize, String> {
    source::set_sources_enabled_batch(source_ids, enabled).map_err(to_cmd_err)
}

#[tauri::command]
pub fn auto_detect_selectors_cmd(
    toc_url: String,
    content_url: String,
    search_url: Option<String>,
) -> Result<crate::spider::DetectedSelectors, String> {
    source::auto_detect_selectors(&toc_url, &content_url, search_url.as_deref())
        .map_err(to_cmd_err)
}

#[tauri::command]
pub fn auto_detect_source_rule(
    name: String,
    toc_url: String,
    content_url: String,
    search_url: Option<String>,
) -> Result<DetectResponse<String>, String> {
    let mut log = DetectLogger::new();
    match source::auto_detect_source_rule_json_with_logs(
        &name,
        &toc_url,
        &content_url,
        search_url.as_deref(),
        &mut log,
    ) {
        Ok(rule_json) => Ok(DetectResponse::ok(rule_json, log)),
        Err(e) => Ok(DetectResponse::fail(e.to_string(), log)),
    }
}

#[tauri::command]
pub fn auto_detect_from_url(url: String) -> Result<DetectResponse<AutoConnectResult>, String> {
    Ok(source::auto_detect_from_url_with_logs(&url))
}

#[tauri::command]
pub fn receive_captured_html(html: String) {
    source::receive_captured_html(html);
}

#[tauri::command]
pub fn auto_detect_from_url_rendered(
    app: tauri::AppHandle,
    url: String,
) -> Result<DetectResponse<AutoConnectResult>, String> {
    Ok(source::auto_detect_from_url_rendered_with_logs(&app, &url))
}
